Fixed hang

Quaker762 opened the pull request:
Fixed Tetris hang.

I'm going to do a complete rework of some of the flags. The flags are the source of all problems...

I hate flags...

wornwinter commented:
What happens now when you run it? Does it progress further than the title screen?

Quaker762 commented:
![w00t](https://cloud.githubusercontent.com/assets/4768900/11032144/7cf8deec-872d-11e5-9610-77cc4fc7ec7e.png)

It gets past the credits but crashes with an uninterrupted exception. The problem was dodgy VBLANK code or weird interrupt handling. The problem now is the register values being all fucked up.

wornwinter commented on src/DMGCPU.cpp:
```diff
@@ -1847,18 +1851,28 @@ void c_DMGCPU::OPCode0x7C()
-    else UNSET_FLAG_BIT(HC_BIT);
-    if((Registers.AF.hi + Registers.HL.lo) > 0xFF) SET_FLAG_BIT(CARRY_BIT);
-    else UNSET_FLAG_BIT(CARRY_BIT);
+    if(result & 0xFF00)
```
Not sure about this. I think that ```if((Registers.AF.hi + Registers.HL.lo) > 0xFF)``` might be a better option, as we need to check if there's a carry from Bit 7.

Quaker762 replied:
Yeah I fixed that up. I'm working on making all the opcodes more consistent, so any old stuff like this/2AM programming will be removed.

wornwinter commented:
I'm nearly sure I made a mistake in BIT 2, b also if you want to look at
that in this pr lol
On 9 Nov 2015 11:02, "Jesse" <notifications@github.com> wrote:

> [image: w00t]
> <https://cloud.githubusercontent.com/assets/4768900/11032144/7cf8deec-872d-11e5-9610-77cc4fc7ec7e.png>
>
> It gets past the credits but crashes with an uninterrupted exception. The
> problem was dodgy VBLANK code or weird interrupt handling. The problem now
> is the register values being all fucked up.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wornwinter/DMGe/pull/36#issuecomment-155029215>.
>

Quaker762 commented:
@wornwinter Okay I'll get on it. I'm going to catch some Z's (I have a Physics exam on Wednesday), so I'll take a look on Thursday unless you want to fix it up :)

wornwinter commented:
I'll get to it tonight lol. Just in work at the minute and can't really
check in depth.
On 9 Nov 2015 11:20, "Jesse" <notifications@github.com> wrote:

> @wornwinter <https://github.com/wornwinter> Okay I'll get on it. I'm
> going to catch some Z's (I have a Physics exam on Wednesday), so I'll take
> a look on Thursday unless you want to fix it up :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wornwinter/DMGe/pull/36#issuecomment-155031977>.
>

Quaker762 commented:
Okay cool. Don't merge by the way I just broke something and it won't load past the Nintendo logo anymore....

Quaker762 commented:
Okay just did a quick check. The problem is at ADD A, (HL). I fucked a flag up. I'll create a PR ASAP

Quaker762 commented:
![wut](https://cloud.githubusercontent.com/assets/4768900/11143723/5dce9eb4-8a4b-11e5-8a52-05302567b8cd.png)
Okay it gets to the menu, but this is the problem I was talking about. 
It sits there for about 10 seconds (keep in mind there's NO cursor/selection arrow displayed) before it does this..

I honestly have no idea what in the hell would cause this... Any ideas??

Quaker762 commented:
![hahaha](https://cloud.githubusercontent.com/assets/4768900/11145017/95e6b520-8a55-11e5-8d4b-065db02d8189.png)

BAHAHAHAHAHA

wornwinter commented:
Haha brilliant. How's it going now?
On 13 Nov 2015 11:26, "Jesse" <notifications@github.com> wrote:

> [image: hahaha]
> <https://cloud.githubusercontent.com/assets/4768900/11145017/95e6b520-8a55-11e5-8d4b-065db02d8189.png>
>
> BAHAHAHAHAHA
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wornwinter/DMGe/pull/36#issuecomment-156402580>.
>

Quaker762 commented:
Pretty good, I added a few more opcodes, however there's still some strange behaviour occuring (it's not crashing to the game screen which is good). Looks like it's getting stuck in a 'cinoop' style loop, so it's possibly something to do with VBLANK.

wornwinter commented:
That's a possibility. The issue with the cursor is likely the lack of
sprites. I don't think I implemented that yet.
On 13 Nov 2015 11:48, "Jesse" <notifications@github.com> wrote:

> Pretty good, I added a few more opcodes, however there's still some
> strange behaviour occuring (it's not crashing to the game screen which is
> good). Looks like it's getting stuck in a 'cinoop' style loop, so it's
> possibly something to do with VBLANK.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wornwinter/DMGe/pull/36#issuecomment-156408039>.
>

Quaker762 commented:
Yeah from the looks of it, its trying to write some OAM stuff (I added some
OAM FMA) to unknown IO registers which could be causing the loop if it's
waiting for return value or something.
On 14 Nov 2015 1:52 am, "Darren Anderson" <notifications@github.com> wrote:

> That's a possibility. The issue with the cursor is likely the lack of
> sprites. I don't think I implemented that yet.
> On 13 Nov 2015 11:48, "Jesse" <notifications@github.com> wrote:
>
> > Pretty good, I added a few more opcodes, however there's still some
> > strange behaviour occuring (it's not crashing to the game screen which is
> > good). Looks like it's getting stuck in a 'cinoop' style loop, so it's
> > possibly something to do with VBLANK.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/wornwinter/DMGe/pull/36#issuecomment-156408039>.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/wornwinter/DMGe/pull/36#issuecomment-156452664>.
>
